### ISSUE TITLE
Form validation issues with ng-messages-include

### ISSUE THREAD

**Issue by jvmccarthy** [angular/material#6810]
I'm trying to start using `ng-messages-include` as described in [ngMessages](https://docs.angularjs.org/api/ngMessages) under "Reusing and Overriding Messages", and I'm running into issues.  I'm primarily seeing it not working with `md-auto-hide`, showing the error messages when the form is displayed.  I like hiding validation errors until the user has had an opportunity to fill out the form and would like to use `md-auto-hide` with `ng-messages-include`. 

See [this CodePen](http://codepen.io/jvmccarthy/pen/EPoywL) for this issue.  The pen is showing other issues including weirdness with adjacent input containers.

Thanks again for all the hard work on angular-material!

**Comment by blaise-io** [comment]
Found the problem:

```js
  angular.forEach(messages.children(), function(child) { 
      // [...]
  });
```

`messages.children()` is wrong because `ng-message`'s aren't necessarily direct children of `ng-messages`. Also, there's [`ng-message-exp`](https://docs.angularjs.org/api/ngMessages/directive/ngMessageExp).

This is wrong in two places: [1](https://github.com/angular/material/blob/1f997951b5d480d04ea7d356107837e279d6a271/src/components/input/input.js#L685), [2](https://github.com/angular/material/blob/1f997951b5d480d04ea7d356107837e279d6a271/src/components/input/input.js#L698).

Solution: It should look for something like `messages.querySelectorAll('ng-message,data-ng-message,x-ng-message,[ng-message],[data-ng-message],[x-ng-message],ng-message-exp,data-ng-message-exp,x-ng-message-exp,[ng-message-exp],[data-ng-message-exp],[x-ng-message-exp]')`. 

That's quite a mouthful, so maybe create a wrapper method similar to [`getMessagesElement`](https://github.com/angular/material/blob/1f997951b5d480d04ea7d356107837e279d6a271/src/components/input/input.js#L744).

**Comment by blaise-io** [comment]
I *nearly* fixed it in https://github.com/blaise-io/material/commit/32b0e7bfac95589e9378c4237a0e5543b90ec533

@topherfangio I noticed you wrote most of this code. Can you help me solve the issue described here? https://github.com/blaise-io/material/commit/32b0e7bfac95589e9378c4237a0e5543b90ec533#diff-da06475ed96460fbcb944d6f1f31748aR605

**Comment by topherfangio** [comment]
@blaise-io Thanks for digging into this. If you can provide a PR with tests, it would be most appreciated!

Regarding the issue you mentioned, you might be able to move that code into a `postLink()` function instead of `compile()` and use `$animate.addClass()` instead of `$element.toggleClass()` to ensure that the animations are properly fired. If that doesn't work, let me know and I can dig in further.

**Comment by blaise-io** [comment]
@topherfangio Thanks for your reply. I tried your suggestion, but the result was that error messages show immediately for both nested `ng-message`s and `ng-message`'s that are direct children. I think it's because when I use a postLink, the `ng-enter` class is missing on the `ng-messages` element, and this class is used to hide the messages initially.

This is what I tried:
```js
function ngMessageDirective($mdUtil, $animate) {
  return {
    restrict: 'EA',
    link: postLink
  };

  function postLink(scope, element) {
    var inputContainer = $mdUtil.getClosest(element, 'md-input-container');

    // If we are not a child of an input container, don't do anything
    if (!inputContainer) return;

    // Add our animation class
    $animate.addClass(element, 'md-input-message-animation');

    return {};
  }
}
```

**Comment by blaise-io** [comment]
By the way I'm wrong about the nesting, `.children()` will actually work as `ng-messages-include` does a replace. But maybe still a good idea to keep the selector in case people use other ways of nesting.

**Comment by DevVersion** [comment]
Not sure, this was covered yet, but the weirdness for the adjacent input containers is caused by a missing HTML tag.

```html
 <div ng-messages-include='error-messages'>
``` 
Should be
```html
 <div ng-messages-include="error-messages"></div>
``` 

About the general issue - will investigate soon.

**Comment by DevVersion** [comment]
FYI: This issue is now fixed in PR #8387
